### PR TITLE
Add SLO table for DataDog Plugin

### PR DESCRIPTION
**Pull request by whume**
# Example query results
Adding Table for SLO's
![image](https://user-images.githubusercontent.com/10459092/200875553-b9e861b2-5357-4ee9-85d3-2cb6246580cd.png)

 Adding an SLO table to allow the user to search SLO's in DataDog and query based on various aspects of the SLO, like targets or timeframes. This is my first time committing a table to this plugin, So if anything needs to be changed more than happy to tweak it. I followed what was set up in the DataDog monitor and plan to continue to add a table for SLO history. 

<details>
  <summary>Results</summary>

  
```
select
  name,
  thresholds
from
  datadog_slo
where
  type in ('monitor') and
  thresholds @> '[{"timeframe":"90d"}]'::jsonb
limit 5;

+-----------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| name                                                | thresholds                                                                                                                                                                                                                                      |
+-----------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Datatech_sql_Host_uptime SLO                        | [{"target":99,"target_display":"99.","timeframe":"90d"}]                                                                                                                                                                                        |
| Kafka: Offline Partitions                           | [{"target":99.9,"target_display":"99.9","timeframe":"7d"},{"target":99.9,"target_display":"99.9","timeframe":"30d"},{"target":99.9,"target_display":"99.9","timeframe":"90d"}]                                                                  |
| Retriever: chatNodeJsErrorLogs by minute (Clone)    | [{"target":99.9,"target_display":"99.9","timeframe":"90d","warning":99.95,"warning_display":"99.95"}]                                                                                                                                           |
| Kafka SLI: Produce request latency                  | [{"target":95,"target_display":"95.","timeframe":"7d","warning":98,"warning_display":"98."},{"target":95,"target_display":"95.","timeframe":"30d","warning":98,"warning_display":"98."},{"target":95,"target_display":"95.","timeframe":"90d"}] |
| Kafka SLI: Consume request latency                  | [{"target":95,"target_display":"95.","timeframe":"7d","warning":98,"warning_display":"98."},{"target":95,"target_display":"95.","timeframe":"30d","warning":98,"warning_display":"98."},{"target":95,"target_display":"95.","timeframe":"90d"}] |
+-----------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
</details>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,10 @@
+## v0.3.0 [2022-11-09]
+
+_What's new?_
+
+- New tables added
+  - [datadog_slo](https://hub.steampipe.io/plugins/turbot/datadog/tables/datadog_slo)
```
Comment (misraved) [inline CHANGELOG.md:6]: Thanks @whume for adding the release note, but it is not necessary to add it as a part of table development. This change can be safely reverted 👍.

File: datadog/plugin.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"datadog_security_monitoring_rule":   tableDatadogSecurityMonitoringRule(ctx),
 			"datadog_security_monitoring_signal": tableDatadogSecurityMonitoringSignal(ctx),
 			"datadog_user":                       tableDatadogUser(ctx),
+			"datadog_slo":                        tableDatadogSLO(ctx),
```
Comment (misraved) [inline datadog/plugin.go:29]: Could you please sort this table list?

Comment (misraved) [inline datadog/plugin.go:29]:
```suggestion
			"datadog_security_monitoring_signal": tableDatadogSecurityMonitoringSignal(ctx),
			"datadog_slo":                                        tableDatadogSLO(ctx),
			"datadog_user":                                     tableDatadogUser(ctx),
```

**Comment by misraved** [comment]
Thanks for creating a PR for adding a new table @whume 👍 . Great to see the contribution 😃 .

I had an initial look at it, and the design looks good 👍. The examples are very sharp and reflect good use cases for the table 👍 .

However, we need to update the table name to `datadog_service_level_objective`. We lean on terraform for table naming convention and would like to keep the deviation to a minimum. Please check https://registry.terraform.io/providers/DataDog/datadog/latest/docs/resources/service_level_objective for additional information.

File: datadog/table_datadog_slo.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+)
+
+func tableDatadogSLO(ctx context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "datadog_slo",
+		Description: "A SLO provides a target percentage of a specific metric over a certain period of time.",
+		List: &plugin.ListConfig{
+			Hydrate: listSLOs,
+			KeyColumns: plugin.KeyColumnSlice{
+				{Name: "name", Require: plugin.Optional},
```
[the remark this answers is not in the document]
Comment (misraved) [reply datadog/table_datadog_slo.go:20]: Only the `id` field is an acceptable optional qual - https://github.com/DataDog/datadog-api-client-go/blob/v1.5.0/api/v1/datadog/docs/ServiceLevelObjectivesApi.md#ListSLOs

```
    optionalParams := datadog.ListSLOsOptionalParameters{
        Ids: &ids,
        Query: &query,
        TagsQuery: &tagsQuery,
        MetricsQuery: &metricsQuery,
        Limit: &limit,
        Offset: &offset,
    }
  ```

File: datadog/table_datadog_slo.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+			// Top columns
+			{Name: "name", Type: proto.ColumnType_STRING, Description: "Name of the SLO."},
+			{Name: "id", Type: proto.ColumnType_STRING, Description: "ID of the SLO."},
+			{Name: "creator_email", Type: proto.ColumnType_STRING, Transform: transform.FromField("Creator.Email"), Description: "Email of the creator."},
+			{Name: "created_at", Type: proto.ColumnType_TIMESTAMP, Transform: transform.FromField("CreatedAt").Transform(convertDatetime), Description: "Timestamp of the SLO creation."},
+			{Name: "type", Type: proto.ColumnType_STRING, Description: "The type of the SLO. For more information about type, see https://docs.datadoghq.com/monitors/service_level_objectives/."},
+
+			// Other useful columns
+			{Name: "modified_at", Type: proto.ColumnType_TIMESTAMP, Transform: transform.FromField("ModifiedAt").Transform(convertDatetime), Description: "Last timestamp when the monitor was edited."},
+			{Name: "monitor_ids", Type: proto.ColumnType_STRING, Description: "The monitor based SLO have monitors assiciated with them. Shows list of associated monitors"},
```
Comment (misraved) [inline datadog/table_datadog_slo.go:33]:
```suggestion
			{Name: "monitor_ids", Type: proto.ColumnType_JSON, Description: "A list of monitor ids that defines the scope of a monitor service level objective."},
```

File: datadog/table_datadog_slo.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+			{Name: "query", Type: proto.ColumnType_STRING, Description: "The Metric based SLOs use querys to detiremine state. Shows associated query"},
```
Comment (misraved) [inline datadog/table_datadog_slo.go:34]:
```suggestion
			{Name: "query", Type: proto.ColumnType_JSON, Description: "The Metric based SLOs use queries to determine the state. Shows associated query."},
```

**Review by misraved — COMMENTED** [review]
I have added a few review comments 👍 . Please take a look at them and let me know if you have questions. Happy to help 😃 .

File: datadog/plugin.go (hunk 2 above)
Comment (whume) [reply datadog/plugin.go]: Sorted

File: docs/tables/datadog_slo.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
+# Table: datadog_slo
+
+<description>
```
[the remark this answers is not in the document]
Comment (whume) [reply docs/tables/datadog_slo.md]: Added missing description

**Comment by whume** [comment]
I added most of the requested changes and appreciate the feedback.  I will look at the GetSLO in a bit and add that in as well. Thanks!

**Comment by whume** [comment]
> On digging deeper, I came across this [API](https://github.com/DataDog/datadog-api-client-go/blob/v1.5.0/api/v1/datadog/docs/ServiceLevelObjectivesApi.md#GetSLO) which can be added easily to include the `GetConfig` configuration. You could easily refer https://github.com/turbot/steampipe-plugin-datadog/blob/main/datadog/table_datadog_dashboard.go#L17 for additional details 👍 .
>
So digging a bit deeper on this could you expand on the value of adding the GET endpoint? I see it's available and it works but it seems to return the same data as the LIST call.  I think long term there is more value tying into the GETSLOHistory API long term but I haven't tested that yet.

**Comment by misraved** [comment]
Thanks for exploring the API @whume 👍. Adding the GET endpoint would enable the user to query the table for a single row. This reduces the number of API calls (if additional ones are added in the future), thereby improving performance.

Details of `GetConfig` can be found here - https://steampipe.io/docs/develop/writing-plugins#hydrate-functions

```
The Get function will usually get the key column data from the QueryData.KeyColumnQuals so that it can get the appropriate item as based on the qualifiers (where clause, join...on). If the Get hydrate function is used as both a Get function AND a normal hydrate function, you should get the key column data from the HydrateData.Item if it is not nil, and use the QueryData.KeyColumnQuals otherwise.
```

**Comment by whume** [comment]
I added the GET endpoint as requested and tested it out there is a new column called `configured_alert_ids` that uses it as that is unique information for that call. Thanks for the explanation very helpful!

**Pull request merged**